wp-includes/mysql/class-mysql-lexer.php

```diff
@@ -0,0 +1,2752 @@
+	}
+
+	protected function number() {
+		$start_position = $this->position;
+
+		if ( '0' === $this->c && 'x' === $this->n ) {
+			$this->hex_number(); // 0x...
+		} elseif ( ( 'x' === $this->c || 'X' === $this->c ) && "'" === $this->n ) {
+			$this->hex_number(); // x'...' or X'...'
+		} elseif ( '0' === $this->c && 'b' === $this->n ) {
+			$this->bin_number(); // 0b...
+		} elseif ( ( 'b' === $this->c || 'B' === $this->c ) && "'" === $this->n ) {
+			$this->bin_number(); // b'...' or B'...'
+		} else {
+			// Here, we have a sequence starting with N or .N, where N is a digit.
+
+			// 1. Try integer first.
+			$this->int_number();
+
+			// 2. In case of N. or .N, it's a decimal or float number.
+			if ( '.' === $this->c ) {
+				$this->consume();
+				$this->type = self::DECIMAL_NUMBER;
+				while ( $this->is_digit( $this->c ) ) {
+					$this->consume();
+				}
+			}
+
+			// 3. When exponent is present, it's a float number.
+			$has_exponent = ( 'e' === $this->c || 'E' === $this->c ) && (
+				$this->is_digit( $this->n )
+				|| ( ( '+' === $this->n || '-' === $this->n ) && $this->is_digit( $this->la( 3 ) ) )
+			);
+			if ( $has_exponent ) {
+				$this->consume(); // Consume the 'e' or 'E'.
+				$this->consume(); // Consume the '+', '-', or digit.
+				while ( $this->is_digit( $this->c ) ) {
+					$this->consume();
+				}
+				$this->type = self::FLOAT_NUMBER;
+			}
+		}
+
+		// In MySQL, when an input matches both a number and an identifier, the number always wins.
+		// However, when the number is followed by a non-numeric identifier-like character, it is
+		// considered an identifier... unless it's a float number, which ignores subsequent input.
+		$possible_identifier_prefix =
+			self::INT_NUMBER === $this->type
+			|| ( '0' === $this->text[0] && ( 'b' === $this->text[1] || 'x' === $this->text[1] ) );
+
+		if ( $possible_identifier_prefix && preg_match( '/\G' . self::PATTERN_UNQUOTED_IDENTIFIER . '/u', $this->input, $matches, 0, $start_position ) ) {
+			$end_position = $start_position + strlen( $matches[0] );
+
+			// When matched more than the number, it's an identifier.
+			if ( $end_position > $this->position ) {
+				$this->text     = $matches[0];
+				$this->position = $end_position;
+				$this->c        = $this->input[ $this->position ] ?? null;
+				$this->n        = $this->input[ $this->position + 1 ] ?? null;
+				$this->type     = self::IDENTIFIER;
+			}
+		}
+	}
+
+	protected function single_quoted_text() {
+		$pattern = $this->is_sql_mode_active( self::SQL_MODE_NO_BACKSLASH_ESCAPES )
+			? self::PATTERN_SINGLE_QUOTED_TEXT_NO_BACKSLASH_ESCAPES
+			: self::PATTERN_SINGLE_QUOTED_TEXT;
+
+		if ( preg_match( '/\G' . $pattern . '/u', $this->input, $matches, 0, $this->position ) ) {
+			$this->text      = $matches[0];
+			$this->position += strlen( $this->text );
+			$this->c         = $this->input[ $this->position ] ?? null;
+			$this->n         = $this->input[ $this->position + 1 ] ?? null;
```

Comment:
`c` and `n` are not meaningful names and require jumping back to the declaration to see what they're about. Let's use human-readable names even if they're a bit longer.